Add AAE fullsync strategy (technology release)

buddhisthead opened the pull request:
This is a squash of branch pdx-aae-fullsync that owes much debt to @jtuple . Thanks, Joe!

The fullsync coordinator is unchanged, while fssource and fssink  get a new capability.
- Source and Sink exchange caps if protocol is >= 2.0.
- If awe is enabled on both sides, they will negotiate the use of AAE strategy
- aae strategy now guarded on existence of cluster aae enabled.
- strategy dynamically determined as each partition is sync'd, so a long sync's strategy can be changed while it's running. If only we had a nice command to do that. For now, you can only change the strategy by changing the app env value in riak_repl of {fullsync_strategy, aee | keylist}.

If a partition is locked by the anti entropy manager, it will be retried periodically until the lock is released, which should only be a few seconds. The 'whereis', 'busy', and reservation mechanisms all still work the same.

For the most part, the new riak_repl_aae_sink and source work identical in their usage to the keylist workers (client and server, respectively).

A PR from riak_kv is required for this PR to complle.
https://github.com/basho/riak_kv/pull/555

And a riak test...
https://github.com/basho/riak_test/pull/282

buddhisthead committed:
Add support for dynamic caps exchange and strategy negotiation.
* Source and Sink exchange caps if protocol is >= 2.0.
* aae strategy now guarded on existence of cluster aae enabled.
* strategy dynamically determined as each partition is sync'd.

buddhisthead committed:
Add support for dynamic caps exchange and strategy negotiation.
* Source and Sink exchange caps if protocol is >= 2.0.
* aae strategy now guarded on existence of cluster aae enabled.
* strategy dynamically determined as each partition is sync'd.

lordnull commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -336,6 +336,7 @@ handle_cast(_Msg, State) ->
 
 %% @hidden
 handle_info({'EXIT', Pid, Cause}, State) when Cause =:= normal; Cause =:= shutdown ->
+    lager:info("fssource ~p exited normally", [Pid]),
```
Shouldn't this be a debug? It's not exactly useful in production.

buddhisthead replied:
Agreed. Left over from debugging.

lordnull commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -383,7 +384,7 @@ handle_info({'EXIT', Pid, Cause}, State) when Cause =:= normal; Cause =:= shutdo
 
 handle_info({'EXIT', Pid, _Cause}, State) ->
-    lager:warning("fssource ~p exited abnormally", [Pid]),
+    lager:info("fssource ~p exited abnormally", [Pid]),
```
Why was this down-graded from a warning?

buddhisthead replied:
You are correct. I should be a warning. Not sure why I changed it. Probably in a flurry of debugging changes.

lordnull commented on src/riak_repl2_fssink.erl:
```diff
@@ -123,16 +196,65 @@ decode_obj_msg(Data) ->
+                    gen_fsm:sync_send_all_state_event(FSW, stop);
+                aae ->
+                    gen_server:call(FSW, stop)
+            end;
```
At some point we should define a fullsync worker behavior so can just call a function rather than case off the strategy and do raw gen_server/gen_fsm messaging.

buddhisthead replied:
Agreed. I considered that and just ran out of time. Calling a function would be better because the implementation (FSM or Gen Server) is exposed here and must change if the implementation of the worker changes. Gross. Sorry.

buddhisthead replied:
Added TODO documentation there to add behavior.

lordnull commented on src/riak_repl2_fssource.erl:
```diff
@@ -54,23 +59,41 @@ init([Partition, IP]) ->
                   {active, false}],
-    ClientSpec = {{fullsync,[{1,1}]}, {TcpOptions, ?MODULE, self()}},
+
+    DefaultStrategy = keylist,
```
Might make more sense for this to be a macro.

lordnull commented on src/riak_repl2_fssource.erl:
```diff
@@ -79,25 +102,64 @@ handle_call({connected, Socket, Transport, _Endpoint, Proto, Props}, _From,
+    Strategy = decide_common_strategy(OurCaps, TheirCaps),
+    lager:info("Common strategy: ~p", [Strategy]),
+
     Transport:setopts(Socket, [{active, once}]),
```
This might be better put inside the "case Strategy of" keylist clause. The socket is given to the aae worker, which means if a message came in between the above {active, once} and the give away, it will come to this process rather than the aae worker.

buddhisthead replied:
But don't we want the {active, once} to take affect for both keylist and aae? I am missing something here.

lordnull replied:
The key difference is in the case of aae strategy, you're giving controll of the socket away, while in keylist you are not. Socket messages when using {active, once} and {active, true} always go to the controlling process. For keylist, that is the same process, but for aae, it is not; thus the potential race condition.

lordnull commented on src/riak_repl2_fssource.erl:
```diff
@@ -79,25 +102,64 @@ handle_call({connected, Socket, Transport, _Endpoint, Proto, Props}, _From,
+            riak_repl_aae_source:start_exchange(FullsyncWorker),
+            {reply, ok,
+             State#state{transport=Transport, socket=Socket, cluster=Cluster,
+                         fullsync_worker=FullsyncWorker, work_dir="/dev/null",
```
Seeing /dev/null is making me cringe. A comment or something on why a flag/atom wouldn't work here?

buddhisthead replied:
Yep. I changed aae's case to use work_dir=undefined and only rm workdir if != undefined on cleanup.

lordnull commented on src/riak_repl_aae_sink.erl:
```diff
@@ -0,0 +1,181 @@
+    RObj = riak_repl_util:from_wire(BObj),
+    %% B = riak_object:bucket(RObj),
+    %% K = riak_object:key(RObj),
+    %% lager:info("PUT ~p:~p", [B, K]),
```
cruft

lordnull commented on src/riak_repl_aae_sink.erl:
```diff
@@ -0,0 +1,181 @@
+
+trace(_Msg) ->
+%%    lager:info("Sink <-------- ~p", [_Msg]),
+    ok.
```
Bascially a no-op, making it crufty.

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+
+    {ok, TreePid} = riak_kv_vnode:hashtree_pid(Partition),
+
+    %% monitor(process, Manager),
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+                   timeout=Timeout,
+                   built=0,
+                   owner=OwnerPid,
+                   wire_ver=w1}, %% can't be w0 because they don't do AAE
```
A note saying this is for future proofing would be nicer. My first reaction was "if it's always the same, then there's no need to put it in the state at all".

buddhisthead replied:
added note re: possible future wire formats

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+    %% exited. Stop exchange.
+    lager:info("Something went down ~p", [Error]),
+    send_complete(State),
+    {stop, something_went_down, State};
```
Store the reference returned from erlang:monitor in the state, pattern match on that, and tell the user what broke.

buddhisthead replied:
Can you refer me to a similar pattern in some other code? I'm not sure what you mean, but it sounds good.

lordnull replied:
riak_repl2_leader uses the described pattern to monitor the leader.

Here's some code for the general pattern as well:

``` erlang
handle_cast({monitor_one, OtherPid}, {_One, Two}) ->
    MonRef = erlang:monitor(process, OtherPid),
    {noreply, {MonRef, Two}};

handle_cast({monitor_two, OtherPid}, {One, _Two}) ->
    MonRef = erlang:monitor(process, OtherPid),
    {noreply, {One, MonRef}}.

handle_info({'DOWN', MonOne, process, Pid, Why}, {MonOne, Two}) ->
    io:formar("Monitored pid one ~p went down due to ~p", [Pid, Why]),
    {noreply, {undefined, Two}};

handle_info({'DOWN', MonTwo, process, Pid, Why}, {One, MonTwo}) ->
    io:format("Monitored pid two ~p went down due to ~p", [Pid, Why]),
    {noreply, {One, undefined}}.
```

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+    lager:info("AAE fullsync source completed partition ~p. Stopping.",
+               [State#state.index]),
+    riak_repl2_fssource:fullsync_complete(Owner),
+%%    {stop, normal, State};
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+                     %% back here once the socket ownership is transfered and
+                     %% we are ready to proceed with the compare.
+                     trace("Init worker_pid", [Partition]),
+                     SourcePid ! {'$aae_src', worker_pid, self()},
```
~~This message never seems to be handled.~~

Handled in compare_loop/1 at the end of the function. Why blocking the fsm, and not handling this in handle_info/3?

buddhisthead replied:
If I understand you correctly, you are suggesting that we instead store Bloom in the state and enter a new state that allows the compare function to run asynchronously, and handle the "done" message with handle_info; and then continue with the process of sending the differences?

This seems like a good idea because it would allow the gen_fsm to return status information during the compare cycle, instead of being blocked.

lordnull replied:
Yes, that sums up what I'm saying.

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+                               end,
+                       lager:info("Full-sync with site ~p; fullsync difference generator for ~p complete (completed in ~p secs)",
+                                  [State#state.cluster, Partition, riak_repl_util:elapsed_secs(StageStart)]),
+                       SourcePid ! {'$aae_src', done, Acc}
```
~~This message never seems to be handled.~~

See earlier comment.

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+    %%         lager:info("Found ~b differences during AAE exchange on ~p of ~p/~p ",
+    %%                    [Count, Cluster, Partition, IndexN])
+    %% end,
+
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+                lager:debug("Keydiff: remote partition ~p remote missing: ~p:~p",
+                            [Partition, Bucket, Key]),
+                ebloom:insert(Bloom, <<Bucket/binary, Key/binary>>),
+                %%ebloom:insert(Bloom, Bin),
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+                {Bucket,Key} = binary_to_term(Bin),
+                lager:debug("Keydiff: remote partition ~p different: ~p:~p",
+                            [Partition, Bucket, Key]),
+                %% ebloom:insert(Bloom, Bin),
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+%%     lager:info("Timeout during exchange between (local) ~p and (remote) ~p, "
+%%                "(preflist) ~p", [LocalVN, RemoteVN, IndexN]),
+%%     send_exchange_status({timeout, RemoteVN, IndexN}, State),
+%%     {stop, normal, State}.
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+%% exchange_complete({LocalIdx, _}, {RemoteIdx, RemoteNode}, IndexN, Repaired) ->
+%%     riak_kv_entropy_info:exchange_complete(LocalIdx, RemoteIdx, IndexN, Repaired),
+%%     rpc:call(RemoteNode, riak_kv_entropy_info, exchange_complete,
+%%              [RemoteIdx, LocalIdx, IndexN, Repaired]).
```
cruft

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+%% @private
+send_exchange_status(Status, State) ->
+    throw(Status),
+    State.
```
send_exhange_status always throws an exception? Why?

buddhisthead replied:
Converted all calls to just throw(xxx). But does a throw cause execution to stop there? Will the next line even be executed?

lordnull replied:
throw/1 halts execution and exits the process with the reason passed in, so the next line is never reached.

http://erlang.org/doc/man/erlang.html#throw-1

```
1> F = fun(A) -> throw(A), io:format("reached it!~n") end.
#Fun<erl_eval.6.82930912>
2> self().
<0.31.0>
3> F(test).
** exception throw: test
4> self().
<0.35.0>
5> catch F(test).
test
6> self().
<0.35.0>
```

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+send_asynchronous_msg(MsgType, #state{transport=Transport,
+                                      socket=Socket}) ->
+    ok = Transport:send(Socket, <<MsgType:8>>).
+
```
while waiting for a reply, any status request will be blocked, as well as cancel messages. It may be better to use gen_tcp:recv/3 (with the timeout) for the blocking receive.

buddhisthead replied:
If I change the code to use a gen_tcp:recv, then I also need to make it use passive mode. And what if it does time out, then what? Either way, we'd be blocking a status call until the timeout is reached. To do this, we'd really have to change how all the messages are threaded and do async io, yes?

lordnull replied:
To do a recv, you would have to put it in passive mode. As for what to do if it times out, either loop around again by sending a 'retry' message to self, or do as the code currently does, and just throw an error. This means the minimum amount of change needed (it remains contained in the function below).

A better way is to do async io, as it would not block status or cancel. Since there are certain messages you expect from the socket, a flag in the state could be used to suss out if it should be handled as a reply, or other data.

buddhisthead replied:
Alright, I added a TODO to use async io; and updated the code to use passive mode with a 60 second timeout. I like this change.

lordnull commented on src/riak_repl_aae_source.erl:
```diff
@@ -0,0 +1,657 @@
+        {Error, Socket, Reason} ->
+            throw({Error, Reason})
+    end.
+
```
trace/1 and trace/2 are basically no-ops. lager:debug/n should likely be used instead.

lordnull commented on src/riak_repl_util.erl:
```diff
@@ -767,7 +767,7 @@ make_pg_name(Remote) ->
 deduce_wire_version_from_proto({_Proto,{CommonMajor,CMinor},{CommonMajor,HMinor}}) ->
     %% if common protocols are both >= 1.1, then we know the new binary wire protocol
-    case CommonMajor >= 1 andalso CMinor >= 1 andalso HMinor >= 1 of
+    case ((CommonMajor >= 2) or (CommonMajor == 1 andalso CMinor >= 1 andalso HMinor >= 1)) of
```
orelse is the short-circuit operator, which would mean the CommonMajor == 1 can be removed.

buddhisthead replied:
I agree that it could be written as ((CommonMajor >= 2) orelse (CMinor >= 1 andalso HMinor >= 1)), but would that make it easier to read? It implies that CommonMajor == 1, but it requires thinking to see it. I like the explicit version better, but that's just my opinion.

lordnull replied:
We could get even crazier and test equivalence directly :) :

``` erlang
deduce_wire_version_from_proto({_Proto,ClientVer,HostVer}) ->
    if
        HostVer >= {1,1} andalso ClientVer >= {1,1} ->
```

My main point was suggesting the short-circuit operator so unneeded work can be skipped.

buddhisthead replied:
Oh, this version is much better. I'm going with that.

buddhisthead committed:
Clean up cruft per Micah's review
* Remove unused code
* Remove noop trace statements
* Repair log messages to original level
* Remove unused variables (including some not from this review)
* Change default strategy to a macro define
* Document some variables

buddhisthead committed:
Clean up cruft per Micah's review
* Remove unused code
* Remove noop trace statements
* Repair log messages to original level
* Remove unused variables (including some not from this review)
* Change default strategy to a macro define
* Document some variables

lordnull committed:
Merge pull request #224 from basho/mw-cascading-rt

Adding cascading realtime. This also adds tests for realtime sinks and sources.

buddhisthead committed:
Move call to set socket options into Keylist strategy stanza.
* This removes the race condition that may have allowed a message
  to be delivered to fssource instead of AAE worker, due to the
  AAE worker not being the controlling process yet.
* Document ownership changes.

buddhisthead committed:
Move call to set socket options into Keylist strategy stanza.
* This removes the race condition that may have allowed a message
  to be delivered to fssource instead of AAE worker, due to the
  AAE worker not being the controlling process yet.
* Document ownership changes.

buddhisthead committed:
Change communications with sink to avoid infinite blocking.
* Changed socket to use passive mode.
* Timeout of 60 seconds on recv from sink.

buddhisthead committed:
Change communications with sink to avoid infinite blocking.
* Changed socket to use passive mode.
* Timeout of 60 seconds on recv from sink.

buddhisthead committed:
Make version comparison even more readable.

buddhisthead commented:
@lordnull There is a problem with the wire format version code. RT and FS are mixing their versions badly. I need to resolve this before we can finish this PR.

buddhisthead committed:
Clean up version detection.
* Remove Version from fullsync source since it doesn't use it.
* RT source helper should use riak_repl_util versioning code.

buddhisthead committed:
Merge branch 'squashed-pdx-aae-fullsync' of github.com:/basho/riak_repl into squashed-pdx-aae-fullsync

Conflicts:
	src/riak_repl2_fssource.erl
	src/riak_repl_util.erl

buddhisthead commented:
@lordnull Ok, I believe I have this back in shape for a final look over. It's rebased against master so it will easy-merge.

lordnull committed:
Removed remnants from a merge

lordnull commented:
I fixed a missed bit from the merge; after that the r_t for this feature passes. Other concerns were addressed, so :+1:

buddhisthead commented:
Many, many thanks to you for the excellent code review, Micah (@lordnull).
